### PR TITLE
Fastpath Write Once Buckets

### PR DESCRIPTION
**Pull request by TheOnlyRew**
https://github.com/basho/internal_wiki/wiki/WriteOnceBuckets

### CODE REVIEW

File: source/languages/en/riak/dev/advanced/write-once.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+possible for users to mis-use the API, accidentally or otherwise, which might
+result in incomparable entries for the same key.
+
+In the case of siblings, write once buckets use an algorithm for resolving
+conflicts that has the appearance of random behavior from the user's point of
+view, but is in fact determinisitic. Effectively, Riak will choose the "least"
+entry among siblings, where the ordering relation is based on a deterministic
+SHA-1 hash of the object. This has the appearance to the user of "random write
+wins", even though the behavior is repeatable and determinisitic (assuming the
+same sibling inputs).
```
Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/write-once.md]: @seemaj made this comment on (the now defunct) eb6ccbefa7e09422bcf2b69592f22e03f99b0836

>I think we should edit this section:
>In the case of siblings, write once buckets use an algorithm for resolving conflicts that has the appearance of random behavior from the user's point of view, but is in fact determinisitic. Effectively, Riak will choose the "least" entry among siblings, where the ordering relation is based on a deterministic SHA-1 hash of the object. This has the appearance to the user of "random write wins", even though the behavior is repeatable and determinisitic (assuming the same sibling inputs).

> to

> In the case of siblings, Riak will choose the "least" entry among siblings, where the ordering relation is based on a deterministic SHA-1 hash of the object.

> I am not sure we should talk about appearing random.

Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/write-once.md]: I disagree with this. Random Write wins is a valid resolution strategy, and that's what developers are going to program around in these cases. The fact that it's actually a deterministic strategy under the hood (That I don't fully understand, honestly. We don't discuss "least" actually means in this context.) doesn't matter unless you have access to the hash operations that make it deterministic --- and user applications won't have that information.

File: source/languages/en/riak/dev/advanced/write-once.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+## Overload Protection
```
Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/write-once.md]: @seemaj made this comment on (the now defunct) eb6ccbe

> We need to remove the section on Overload Protection. Talked to Fred and he agreed that we should not advertise this at the moment. The way the things are now, sometimes users will get an error and sometimes the node will crash. There is a ticket to fix this.

Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/write-once.md]: Should we make users aware of this deficiency in some other way?

File: source/languages/en/riak/dev/advanced/bucket-types.md [hunk 3]
```diff
@@ -37,9 +37,9 @@ with a few crucial differences:
   those configurations to as many buckets as you wish, whereas the
   previous system required configuration to be set on a per-bucket basis
 * Nearly all bucket properties can be updated using bucket types, with
-  two exceptions: the `datatype` and `consistent` properties related to
-  [[Riak Data Types|Data Types]] and [[strong consistency]],
-  respectively
+  three exceptions: the `datatype`, `consistent`, and `write_once`
+  properties related to [[Riak Data Types|Data Types]],
+  [[strong consistency]], and [[Write Once Buckets]] respectively
```
Comment (laurenrother) [inline source/languages/en/riak/dev/advanced/bucket-types.md]: I would edit this sentence to read: 
"Nearly all bucket properties can be updated using bucket types, except the `datatype`, `consistent`, and `write once` properties related to..."

File: source/languages/en/riak/dev/advanced/bucket-types.md [hunk 4]
```diff
@@ -82,6 +82,9 @@ system of bucket configuration, including the following:
   `riak-admin bucket-type` interface (discussed in depth below) enables
   you to manage bucket configurations on the operations side, without
   recourse to Riak clients.
+* Some special usecases -- [[Strong Consistency|Managing Strong Consistency]],
+  [[Data Types|Using Data Types]], and [[Write Once Buckets]] -- are only
+  available through bucket properties, or bucket types.
```
Comment (laurenrother) [inline source/languages/en/riak/dev/advanced/bucket-types.md]: No comma needed before 'or'. :)

File: source/languages/en/riak/dev/advanced/write-once.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
+---
+title: Write Once Buckets
+project: riak
+version: 2.1.0+
+document: tutorial
+audience: intermediate
+keywords: [developers, buckets]
+---
+
+Riak 2.1.0 introduces the concept of "write once" buckets; buckets whose entries
```
Comment (laurenrother) [inline source/languages/en/riak/dev/advanced/write-once.md]: Can we change this to:
"Riak 2.1.0 introduces the concept of Write Once Buckets, buckets whose entries are intended to be written exactly once and never updated or overwritten."

File: source/languages/en/riak/dev/advanced/write-once.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+are intended to be written exactly once, and never updated or over-written.
+Buckets of this type have the special feature that they circumvent the normal
+"coordinated put" path, which would otherwise result in a read on the
+coordinating vnode before the write. Avoiding coordinated puts results in higher
```
Comment (laurenrother) [inline source/languages/en/riak/dev/advanced/write-once.md]: I would edit this sentence to read: 

"Buckets of this type circumvent the normal "coordinated put" path..." 
Is it put or PUT?

File: source/languages/en/riak/dev/advanced/write-once.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+throughput and lower put latency, at the cost of different semantics in the
+degenerate case of sibling resolution.
+
+## Configuration
+
+We introduce the `write_once` bucket type property, which, when set to `true`,
+treats all key/value entries as semantically "write once"; once written, entries
+should not be modified or over-written by the user.
```
Comment (laurenrother) [inline source/languages/en/riak/dev/advanced/write-once.md]: When the `write_once` bucket type property is set to `true`, it treats all key/value entries as semantically "write once"; once written, entries should not be modified or overwritten by the user.

File: source/languages/en/riak/dev/advanced/write-once.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+The `write_once` property is a boolean property that is applied to a bucket type
+and may only be set at bucket creation time. Once a bucket type has been set
```
Comment (laurenrother) [inline source/languages/en/riak/dev/advanced/write-once.md]: Remove the 'that is'.

File: source/languages/en/riak/dev/advanced/write-once.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+prompt$ riak-admin bucket-type status my-bucket-type
+my-bucket-type is active
+...
+write_once: true
+...
+```
+
+## Quorum
+
+The write once path supports the `W`, `PW`, and `DW` configuraton values.
```
Comment (laurenrother) [inline source/languages/en/riak/dev/advanced/write-once.md]: So write once in this section is lower case with no ` `. And below, in ##Runtime, it's back in " ". I think to be consistent on the page, it needs to be capitalized or put in backticks if we usually backtick paths.

File: source/languages/en/riak/dev/advanced/write-once.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+## Runtime
+
+The "write once" path circumvents the normal coordinated put code path, and
+instead sends write requests directly to all vnodes (or vnode proxies) in the
+effective preference list for the write operation.
+
+In place of the `put_fsm` used in the normal path, we introduce a collection of
+new intermediate worker processes (implementing `gen_server` behavior). The role
+of these intermediate processes is to dispatch put requests to vnode(proxie)s in
+the preflist, and to aggregate replies. Unlike the `put_fsm`, the write once
```
Comment (laurenrother) [inline source/languages/en/riak/dev/advanced/write-once.md]: No comma before 'and'. :)

File: source/languages/en/riak/dev/advanced/write-once.md [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+workers are long-lived for the lifecyle of the `riak_kv` application. They are
+therefore stateful, and store request state in a state-local dictionary.
```
Comment (laurenrother) [inline source/languages/en/riak/dev/advanced/write-once.md]: No comma needed before 'and' here either.

File: source/languages/en/riak/dev/advanced/write-once.md [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+The relationship between the `riak_client`, write once workers, and vnode
+proxies is illustrated in the following diagram:
+
+<br>
+![Write Once](/images/write_once.png)
+<br>
+
+## Siblings
+
+As mentioned, entries in write once buckets are intended to be written only once
```
Comment (laurenrother) [inline source/languages/en/riak/dev/advanced/write-once.md]: Write Once Buckets

File: source/languages/en/riak/dev/advanced/write-once.md (hunk 1 above)
Comment (laurenrother) [inline source/languages/en/riak/dev/advanced/write-once.md]: Is the deterministic SHA-1 hash something the users can control/influence/impact? If so, then I agree that talking about it appearing to be random write is not as helpful. Instead, we could maybe suggest ways the user can impact the outcome.

If they can't impact the outcome, though, I would maybe rephrase to something like, 
"In the case of siblings, Write Once Buckets uses a deterministic algorithm based on SHA-1 hashes to resolve conflicts. While Riak uses the "least" entry among siblings, you do not have much control over which entry is least so the result may appear to be "random write wins"."

Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/write-once.md]: Splitting the difference. Changing to 

```
In the case of siblings, Write Once Buckets will resolve the conflict by
choosing the "least" entry, where sibling ordering is based on a deterministic
SHA-1 hash of the objects. While this algorithm is repeatable and deterministic
at the database level, it will have the appearance to user of "random write
wins".
```

File: source/languages/en/riak/dev/advanced/bucket-types.md (hunk 3 above)
Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/bucket-types.md]: Done.

File: source/languages/en/riak/dev/advanced/bucket-types.md (hunk 4 above)
Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/bucket-types.md]: Removed.

File: source/languages/en/riak/dev/advanced/write-once.md (hunk 5 above)
Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/write-once.md]: Changed to
```
Riak 2.1.0 introduces the concept of Write Once Buckets, buckets whose entries
are intended to be written exactly once, and never updated or over-written.
```

File: source/languages/en/riak/dev/advanced/write-once.md (hunk 6 above)
Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/write-once.md]: Going with

```
Buckets of this type circumvent the normal "coordinated PUT" path . . .
```

File: source/languages/en/riak/dev/advanced/write-once.md (hunk 7 above)
Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/write-once.md]: Changed to

```
When the new `write_once` bucket type property is set to `true`, buckets using
that type will treat all key/value entries as semantically "write once"; once
written, entries should not be modified or overwritten by the user.
```

File: source/languages/en/riak/dev/advanced/write-once.md (hunk 8 above)
Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/write-once.md]: ```The `write_once` property is a boolean property applied to a bucket type and may```

File: source/languages/en/riak/dev/advanced/write-once.md (hunk 9 above)
Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/write-once.md]: Changed here to 

```
The write path used by Write Once Buckets supports the `W`, `PW`, and `DW`
configuration values. However, if `DW` is specified, then the value of `W` is
taken to be the maximum of the `W` and `DW` values. For example, for `n_val 3`,
if `DW` is set to `all`, then `W` will be `3`.

This write additionally supports the `sloppy_quroum` property. If set to
`false`, only primary nodes will be selected for calculation of write quorum
nodes.
```

I'll do my best to remove instances of `write once` and `"write once"`, so that we're left with the Strong Case concept of `Write Once Buckets`, and the actual `"\`write_once`\" configuration option.

File: source/languages/en/riak/dev/advanced/write-once.md (hunk 10 above)
Comment (TheOnlyRew) [inline source/languages/en/riak/dev/advanced/write-once.md]: Fixed.

**Comment by TheOnlyRew** [comment]
334c787 was pushed so that @laurenrother could easily see/comment on the changes made in response to her comments on dca3e50.
I intend to squash the two commits and force-push the new history before we merge this PR, though.

**Comment by TheOnlyRew** [comment]
This PR was locally cherry-picked as part of a re-write of the history for PR #1625---specifically 6d47f89.

This PR will be closed, and the branch deleted.

File: source/languages/en/global_nav.yml [hunk 13]
```diff
@@ -92,6 +92,7 @@ riak:
           - "[[Commit Hooks|Advanced Commit Hooks]]"
           - "[[Replication Properties]]"
           - "[[Strong Consistency|Using Strong Consistency]]"
+          - "[[Write Once Buckets]]"
```
Comment (lucperkins) [inline source/languages/en/global_nav.yml]: My completely unwanted $.02 here is to make the official term "write-once buckets" and, correspondingly, to title the doc "Write-once Buckets" (in accordance with Basho's capitalization/hyphenation standards).

Aaaaaand I'm out.
